crates/sui-indexer/migrations/pg/2024-09-12-213234_watermarks/up.sql

```diff
@@ -0,0 +1,34 @@
+CREATE TABLE watermarks
+(
+    -- The table name of group of tables governed by this watermark, i.e `epochs`, `checkpoints`,
+    -- `transactions`. For example, `transactions` governs the `transactions` table and associated
+    -- lookup tables.
+    entity                      TEXT          NOT NULL,
+    -- Inclusive upper bound epoch this entity has data for. Committer updates this field. Pruner
+    -- uses this field for per-entity epoch-level retention, and is mostly useful for pruning
+    -- unpartitioned tables.
+    epoch_hi                    BIGINT        NOT NULL,
+    -- Inclusive lower bound epoch this entity has data for. Pruner updates this field, and uses
+    -- this field in tandem with `epoch_hi` for per-entity epoch-level retention. This is mostly
+    -- useful for pruning unpartitioned tables.
+    epoch_lo                    BIGINT        NOT NULL,
```

Comment:
Thanks for adding the explanation here @wlmyng -- I'm still a little lost on why unpartitioned tables need epoch bounds on top of their existing `hi` and `lo` columns though, could you explain that too?

Reply:
a single entry like `transactions` has `epoch_lo` and `epoch_hi` which govern epoch-partitioned tables, and `lo`, `pruned_lo`, and `hi` which are used for the lookup tables. Epoch-partitioned tables will prune between `[min_partition..epoch_lo)`, while lookup tables prune between `(pruned_lo, lo)`

Reply:
Revisiting this. Now that `epochs` table is no longer pruned, and considering the latest updates to the watermarks table design, I think we can get rid of `epoch_lo` and `epoch_hi` after all.

The latest epoch from `epochs` will inform the pruner what the current epoch is, and for each prunable table, the `epoch_lo` can be easily calculated. If data exceeds the retention policy for the table, it can just consult the `epochs` table for the corresponding unit for the table (tx or cp seq num)